### PR TITLE
Testing readme

### PR DESCRIPTION
**Pull request by sasmith**
Add a tests/README.md. Intended for consumption by LLMs so they know where to write tests, and by humans so they also know about tests.

### CODE REVIEW

**Comment by sasmith** [comment]
* **#253** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/253?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/253?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: tests/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Testing Philosophy
+
+Follow generic best practices. With help from our friend Claude, here are some generic points
+* Tests should be independent and idempotent. Tests shouldn't pass or fail based on other tests.
+* Tests should be focussed. Test one thing per test.
```
Comment (jacklionheart) [inline tests/README.md:5]: s/focussed/focused

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+* Tests should be fast and efficient.
+* Tests should cover edge cases and boundary conditions, not just the happy path.
+* External dependencies should be appropriately mocked, so your test can be focused on what it's testing.
+* Tests should be readable by humans.
```
Comment (jacklionheart) [inline tests/README.md:9]: I will take a task to give this a slight massage. I think there is a counterbalance to the "edge cases" point which is also about "defining the core behaviors of the file/class/process being tested and testing exactly those, not implementation details thereof"

File: tests/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+In addition to these generic points, here are some other thoughts:
+* If you need to fix a bug or a regression, you should probably write a test that would have caught it.
```
Comment (jacklionheart) [inline tests/README.md:12]: I might say "first try to reproduce it as a unit test"

File: tests/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+* Unit tests are better because they're faster. Integration tests are better because they're more comprehensive.
```
Comment (jacklionheart) [inline tests/README.md:13]: I don't think integration tests are *comprehensive*, theyre *directly meaningful* in that they can test what we actually care most directly about (e.g. training runs all the way through)

File: tests/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+* LLMs should be writing most of your tests, but you should review them to make sure they're covering what they should be.
```
Comment (jacklionheart) [inline tests/README.md:14]: There might be something to say about tests being "disposable"

**Review by jacklionheart — APPROVED** [review]

File: tests/README.md (hunk 1 above)
Comment (sasmith) [reply tests/README.md:5]: I was trying to add colour.

File: tests/README.md (hunk 5 above)
Comment (sasmith) [reply tests/README.md:14]: Feel free to adjust when you take a pass at this.

**Comment by sasmith** [comment]
Update in response to about half of your comments. Happy for you to take another pass. The main reason I wanted to add this was for agents to be able to read about the structure (I have only a limited guess about how much this matters, but happy to learn more). The rest felt like ... the readme would be more empty without saying more on testing philosophy? But I didn't want to get hung up on it, since I think that 90% of it is "you know, good testing practices".

File: tests/README.md (hunk 1 above)
Comment (jacklionheart) [reply tests/README.md]: look we both know i can generate enough typos for the both of and I don't really like you stepping on my AoRs
